### ISSUE TITLE
Slash encoding

### ISSUE THREAD

**Issue by theatrain** [angular-ui/ui-router#1645]
I am still getting the problem when trying to pass dates.

    ui-sref="jobs-list({quote_due_from:'18/12/2014',quote_due_to:'18/12/2014'})"

The following URL is generated:

    http://{mydomain}/#/jobs?quote_due_from=18%252F12%252F2014&quote_due_to=18%252F12%252F2014

This is my state provider code:

    module.config(function($stateProvider) {
        $stateProvider
            .state('jobs-list', {
                url: '/jobs?quote_due_from&quote_due_to&quote_followup_from&quote_followup_to&start_date_from&start_date_to&quote_status_id',
                controller: 'JobsListCtrl',
                templateUrl: 'jobs/jobs.list.html'
            });
    });

I am using the latest version of ui.router

    "name": "angular-ui-router",
    "version": "0.2.13",

**Comment by kcjmowright** [comment]
I got around this issue by explicitly configuring the parameter types to 'any'.

```
module.config(function($stateProvider) {
    $stateProvider
        .state('jobs-list', {
            url: '/jobs?{quote_due_from:any}&{quote_due_to:any}&{quote_followup_from:any}&{quote_followup_to:any}&{start_date_from:any}&{start_date_to:any}&quote_status_id',
            controller: 'JobsListCtrl',
            templateUrl: 'jobs/jobs.list.html'
        });
});
```

**Comment by tweinerpq** [comment]
Hi,
I'm still experiencing double encoding when the string I'm sending in is part of the URL, like so:
example.com/id

I'm sending in this:
<a ui-sref="search.doc({ id: 'CBVPPYMD10.1017/S003329171400227X' })" ...

and on click I'm seeing on my browser

url.com/editor/CBVPPYMD10.1017%252FS003329171400227X

I'm also seeing this same doubly encoded string as the href attribute generated by ui-router on Inspect Element.

My real doc id is CBVPPYMD10.1017/S003329171400227X

I just updated to @version v0.2.15 and I have latest angular (1.4)

Thanks.

**Comment by tweinerpq** [comment]
Oh, I see that this is being addressed here?
https://github.com/angular-ui/ui-router/pull/2071

Is this supposed to be fixed?

**Comment by chliebel** [comment]
@tweinerpq #2071 won’t change the “double encoding” of the slash in the URL, however it will fix the decoding, so the encoded slash from the URL (`%252F`) is correctly decoded to the slash character (`/`) in the state params, as soon as the PR is merged.

**Comment by jshado1** [comment]
:+1:

The problem comes from [UrlMatcherFactory.js:360](https://github.com/angular-ui/ui-router/blob/master/src/urlMatcherFactory.js#L360). The value is encoded on line 338, and then that encoded value is passed to encodeURIComponent on line 360:

```js
/* 338 */ var encoded = param.type.encode(value);
// …
/* 360 */ encoded = map(encoded, encodeURIComponent).join('&' + name + '=');
```

**Comment by zgmnkv** [comment]
I have made some research on this.
Actual problem is in valToString function [urlMatcherFactory.js:572](https://github.com/angular-ui/ui-router/blob/master/src/urlMatcherFactory.js#L572)
```JavaScript
  function valToString(val) { return val != null ? val.toString().replace(/\//g, "%2F") : val; }
  function valFromString(val) { return val != null ? val.toString().replace(/%2F/g, "/") : val; }

  var $types = {}, enqueue = true, typeQueue = [], injector, defaultTypes = {
    string: {
      encode: valToString,
      decode: valFromString,
```

This behavior was introduced in 3045e41577a8b8b8afc6039f42adddf5f3c061ec and 0c983a08e2947f999683571477debd73038e95cf fixing #1119 
Seems that it fixes path params, but breaks search params
Maybe we should encode/decode path params and search params differently

**Comment by zivc** [comment]
Solution here:
https://github.com/angular-ui/ui-router/issues/1119#issuecomment-134603427

**Comment by mismith** [comment]
Not in Chrome, @zivc 
See: https://github.com/angular-ui/ui-router/issues/1119#issuecomment-134686629

**Comment by frogomatic** [comment]
Is there any reason why one would ever want to encode the slash (via valToString) and then encodeURIComponent on top of that, either in parameters or in a path component?

**Comment by leye0** [comment]
What I don't like is that the href is shown with an encoded slash when passing a url with a slash as a parameter to ui-sref: site.com/page%2Fhome. I feel like this is not good when search engines are crawling the website for links.

**Comment by gligoran** [comment]
Any update on this bug? It's causing me a few problems. It's not only that the slashes are double-encoded, but it make decoding them wrong, because by manually replacing those slashes you take them out of step with the rest of encoding/decoding.

For example let's say this is URL: ```http://example.com/q=/--+```

The proper way to encode it would be: ```http://example.com/q=%2f--%2b```

But because of the manual slash encoding by uiRouter and then usage of ```encodeURIComponent``` the resulting URL is this: ```http://example.com/q=%252f--%2b```

Now when you try to decode the ```q``` parameter you get ```%2f--+``` and if you try to do it again, you get ```/-- ``` (the last character is a space).

This might be a simple example, but in my case this is causing me a lot of problems with email confirmation tokens.

**Comment by nguillaumin** [comment]
We are experiencing a similar problem: One of the parameter in our routes is a URL, e.g. `/something/:url` meaning that it needs to be properly encoded for it to work. We spent some time investigating that and we have I think a good grasp of it so I'll dump our analysis here in the hope it will be helpful in finding a solution.

Pre-requisite reading: http://blog.lunatech.com/2009/02/03/what-every-web-developer-must-know-about-url-encoding

I think the core of the problem is that we need to consider that we're inside the *fragment* part of the browser URL, e.g. `http://example.com/angular-app/#/this-is-the-fragment-part`. Values in the fragment part are encoded in a specific way where the following characters are permitted *unencoded*: `/?:@-._~!$&'()*+,;=`

On a regular (non AngularJS) link, it means you can have a valid fragment part like `http://example.com/page.html#http://www.example.com/?key=value` .

But within an AngularJS app you can't have that, because `/` is used to separate route parts `http://www.example.com/angular-app/#/route/child-route/grand-child-route`. If the `/` were left unencoded you could have `http://www.example.com/angular-app/#/route/http://example.com/something.html` which would result in the route parts:

* `route`
* `http:`
* `[empty]`
* `example.com`
* `something.html`

What you really want is: `http://www.example.com/angular-app/#/route/http:%2F%2Fexample.com%2Fsomething.html` to get only 2 route parts `route` and `http://example.com/something.html`.

So I suspect the solution is to encode/decode the values following the rules for [URL fragment parts](https://tools.ietf.org/html/rfc3986#section-3.5) but to make an exception just for slashes. Not respecting the fragment rules may cause problems with other special chars ``/?:@-._~!$&'()*+,;=` . I think only the slash is special in our case.

To add to the confusion, `encodeURI()` and `encoderURIComponent()` cannot be used as-is to do this encoding because they are too aggressive and not suited to encode fragment parts. It seems to be a known problem by the AngularJS developers who come up with their own functions to do so:

* https://github.com/angular/angular.js/blob/e57cf13d5da204a3324e8b083718c0dcfbb58207/src/Angular.js#L1328
* https://github.com/angular/angular.js/blob/e57cf13d5da204a3324e8b083718c0dcfbb58207/src/Angular.js#L1347

Hope this helps!

**Comment by schneikai** [comment]
Same issue here. Slashes in query string parameters are getting double encoded when going through *$state.go* for example.

*C/L* becomes *C%252FL* where it should be *C%2FL*

**Comment by christopherthielen** [comment]
This should be fixed in release 0.2.16+ We now encode slashes specially, using a custom escape character (tilde) 
see  02e9866

**Comment by Red-3** [comment]
I have a problem since upgrading ui-router in an app which is using ui-router's HTML5 mode. The path part of the URL is getting encoded with ~2F. Is this because the code is incorrectly encoding URLs parsed in HTML5 mode?
From the discussion above it appears to me that only the _fragment part_ of the URL should be encoded with this change. That is clearly not happening in my case.

Since upgrading ui-router the URL after a state.go() is rewritten in the browser as this:
https://localhost:8443/~2Fdocp~2Fapp~2Fwelcome~2Findex.html/register?access_token=4c1b26d8-bd1b-41b4-b7a6-3d7c748b3235
(Note that it only encodes _some_ of the forward slashes.)
It should be:
https://localhost:8443/docp/app/welcome/index.html/register?access_token=4c1b26d8-bd1b-41b4-b7a6-3d7c748b3235

I have tested other apps after upgrading ui-router _not_ using HTML5 mode and the URLs are written (as expected) without the ~2F encoding.

Should I raise a new issue for this?

**Comment by fallXone** [comment]
Seems like this just makes an error less likely to happen by doing your own encoding with ~, while introducing another issue. What happens if ~2F is in a parameter? Then it will get decoded to a /. How come you guys aren't using encodeURIComponent?

**Comment by christopherthielen** [comment]
> How come you guys aren't using encodeURIComponent?

@fallXone because angular $location has special handling for splitting on '/' which causes encoded parameter values with slashes like `/foo/param%2Fval/bar` to be forcibly decoded to `/foo/param/val/bar`.

See my analysis ticket: https://github.com/angular-ui/ui-router/issues/2598

**Comment by abhijeetahuja** [comment]
url: 'products/{slugPath:.*}'
with
url: 'products/{slugPath:any}'

**Comment by jsheffers** [comment]
Angular 1.6.7 introduced a change of how forward slashes are handled.

https://github.com/angular/angular.js/commit/e06ebfdbb558544602fe9da4d7d98045a965f468

https://github.com/angular/angular.js/issues/16312

This broke our current angular app which is using url: '{aliasUrl:any}'.  Has this change broken any other apps?  Temporarily we've reverted back to 1.6.6 to resolve the issue.

**Comment by albertvaka** [comment]
Broke our app too, URLs get double-encoded (%20 becomes %2520).

**Comment by alee046** [comment]
Does anyone have any updates on this issue? It is breaking our site as well.

**Comment by christopherthielen** [comment]
Follow ther angular repository for updates on this.

Related angularjs 1.6.7 issues and prs:

- https://github.com/angular/angular.js/issues/16312
- https://github.com/angular/angular.js/pull/16316
- https://github.com/angular/angular.js/pull/16350
- https://github.com/angular/angular.js/pull/16354

**Comment by albertvaka** [comment]
So, has this been fixed in 1.6.8?

**Comment by christopherthielen** [comment]
The uirouter/angularjs test suite passes using angular 1.6.8 (it does not pass with 1.6.7)
